### PR TITLE
ENT-5273 unified database management

### PR DESCRIPTION
**Pull request by blsemo**
This PR delivers the epic ENT-5273 - harmonising database/schema management between OS and ENT and making schema set-up a separate step when setting up or upgrading a node.

### CODE REVIEW

**Comment by blsemo** [comment]
Run tests: slow, smoke

File: core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt [hunk 1]
```diff
@@ -116,7 +116,7 @@ class FlowIsKilledTest {
     }
 
     @Test(timeout = 300_000)
-    fun `manually handle killed flows using checkFlowIsNotKilled`() {
+    fun `manually handle killed flows using checkForIsNotKilled`() {
```
Comment (rick-r3) [inline core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt]: Why the rename here ?

File: core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt [hunk 2]
```diff
@@ -135,7 +135,7 @@ class FlowIsKilledTest {
     }
 
     @Test(timeout = 300_000)
-    fun `manually handle killed flows using checkFlowIsNotKilled with lazy message`() {
+    fun `manually handle killed flows using checkForIsNotKilled with lazy message`() {
```
Comment (rick-r3) [inline core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt]: Why the rename here ?

**Review by mnesbit — COMMENTED** [review]
I certainly don't see anything here to worry about. We only have judge the most practical testing strategy and how we merge/update all the downstream dependencies

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt [hunk 3]
```diff
@@ -26,22 +25,19 @@ abstract class BaseSessionFactoryFactory : CordaSessionFactoryFactory {
...
-                } else {
-                    "none"
                 }
         // We set a connection provider as the auto schema generation requires it.  The auto schema generation will not
         // necessarily remain and would likely be replaced by something like Liquibase.  For now it is very convenient though.
         return Configuration(metadataSources).setProperty("hibernate.connection.provider_class", HibernateConfiguration.NodeDatabaseConnectionProvider::class.java.name)
                 .setProperty("hibernate.format_sql", "true")
                 .setProperty("javax.persistence.validation.mode", "none")
-                .setProperty("hibernate.connection.isolation", databaseConfig.transactionIsolationLevel.jdbcValue.toString())
+                .setProperty("hibernate.connection.isolation", TransactionIsolationLevel.default.jdbcValue.toString())
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt:40]: This looks a bit like moving from a value in the config to a default.  Is this correct?  Does it matter?

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt [hunk 4]
```diff
@@ -48,6 +48,20 @@ open class SharedNodeCmdLineOptions {
...
 
+    @Option(
+            names = ["--allow-hibernate-to-manage-app-schema"],
+            description = ["Allows hibernate to create/modify app schema for CorDapps based on their mapped schema.",
+                "Use this for rapid app development or for compatibility with pre-4.6 CorDapps.",
+                "Only available in dev mode."]
+    )
+    var allowHibernateToManangeAppSchema: Boolean = false
+
+    @Option(
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt]: Why is this change here ?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt (hunk 3 above)
Comment (mnesbit) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt:40]: Yes this is intentionally breaking so that people will delete the 'isolation' level from their DB config. We considered ignoring any config, but felt that this would give people a false sense of how things are setup.

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt (hunk 4 above)
Comment (blsemo) [reply node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt]: This allows you to run in a mode compatible with the old behaviour (i.e. the node will set up app schema on the fly via hibernate).

File: core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt (hunk 1 above)
Comment (blsemo) [reply core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt]: No idea - that came in via some merge from someone else. I can go in and revert it.

Comment (blsemo) [reply core-tests/src/test/kotlin/net/corda/coretests/flows/FlowIsKilledTest.kt]: I added a commit to the latest "update from release branch" PR that will revert this change.

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt (hunk 4 above)
Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt]: No, no.   This command line option for pausing flows.  Looks like bad merge ?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt (hunk 3 above)
Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt:40]: Okay....  A couple of points:

1. The isolation level is still configurable.  This config element was always redundant because the data source allows it to be configured.  e.g. see `NodeConfigurationImpl`

```
if (dataSourceProperties["transactionIsolation"] == null) {
    dataSourceProperties["transactionIsolation"] = database.transactionIsolationLevel.jdbcString
}
```

2. Our docs say to use READ_COMMITTED with all the databases (https://docs.corda.net/docs/corda-enterprise/4.4/node/operating/node-database-developer.html#sql-server-1).  So we should change the default from the one in this PR.

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt (hunk 4 above)
Comment (blsemo) [reply node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt]: Interesting. That got added to the release branch and then pulled in here and then removed again from the release branch (but the removal didn't make it here).

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt (hunk 3 above)
Comment (blsemo) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt:40]: Well, we have to decide whether our previous code or the docs are right - the default level in (`reference.conf` and the default value in the code) are `REPEATABLE_READ` - no one has used `READ_COMMITTED` for the last 2 years, I guess...

Comment (blsemo) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/factory/BaseSessionFactoryFactory.kt:40]: 🤦 We have been using `READ_COMMITTED` in the perf cluster and told everyone to use it in the docs but our default was just wrong.

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt [hunk 5]
```diff
@@ -48,6 +48,14 @@ open class SharedNodeCmdLineOptions {
     )
     var devMode: Boolean? = null
 
+    @Option(
+            names = ["--allow-hibernate-to-manage-app-schema"],
+            description = ["Allows hibernate to create/modify app schema for CorDapps based on their mapped schema.",
+                "Use this for rapid app development or for compatibility with pre-4.6 CorDapps.",
+                "Only available in dev mode."]
+    )
+    var allowHibernateToManangeAppSchema: Boolean = false
```
Comment (vkolomeyko) [inline node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt]: Spelling: `allowHibernateToManageAppSchema`

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt [hunk 6]
```diff
@@ -48,6 +48,14 @@ open class SharedNodeCmdLineOptions {
     )
     var devMode: Boolean? = null
 
+    @Option(
+            names = ["--allow-hibernate-to-manage-app-schema"],
+            description = ["Allows hibernate to create/modify app schema for CorDapps based on their mapped schema.",
+                "Use this for rapid app development or for compatibility with pre-4.6 CorDapps.",
+                "Only available in dev mode."]
```
Comment (vkolomeyko) [inline node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt:55]: Where is it enforced, please?
I.e. if I start in `devMode = false` what would stop me passing `--allow-hibernate-to-manage-app-schema` ?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt [hunk 7]
```diff
@@ -64,10 +65,10 @@ class HibernateConfiguration(
     fun sessionFactoryForSchemas(key: Set<MappedSchema>): SessionFactory = sessionFactories.get(key, ::makeSessionFactoryForSchemas)!!
 
     private fun makeSessionFactoryForSchemas(schemas: Set<MappedSchema>): SessionFactory {
-        val sessionFactory = sessionFactoryFactory.makeSessionFactoryForSchemas(databaseConfig, schemas, customClassLoader, attributeConverters)
+        val sessionFactory = sessionFactoryFactory.makeSessionFactoryForSchemas(schemas, customClassLoader, attributeConverters, allowHibernateToManageAppSchema)
 
         // export Hibernate JMX statistics
-        if (databaseConfig.exportHibernateJMXStatistics)
+        if (exportHibernateJMXStatistics)
```
Comment (vkolomeyko) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/HibernateConfiguration.kt:71]: Method `initStatistics` can be made `private`

File: testing/smoke-test-utils/src/main/kotlin/net/corda/smoketesting/NodeProcess.kt [hunk 8]
```diff
@@ -138,9 +143,25 @@ class NodeProcess(
             }
         }
 
-        private fun startNode(nodeDir: Path): Process {
+        class SchemaCreationTimedOutError(nodeDir: Path) : Exception("Creating node schema timed out for $nodeDir")
+        class SchemaCreationFailedError(nodeDir: Path) : Exception("Creating node schema failed for $nodeDir")
+
+
+        private fun createSchema(nodeDir: Path){
+            val process = startNode(nodeDir, arrayOf("run-migration-scripts", "--core-schemas", "--app-schemas"))
```
Comment (vkolomeyko) [inline testing/smoke-test-utils/src/main/kotlin/net/corda/smoketesting/NodeProcess.kt:151]: Looks like we have a duplicated array constant in: `net.corda.demobench.model.NodeController#schemaSetupArgs`
Would it make sense to refactor and use it from both places?

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt [hunk 9]
```diff
@@ -75,6 +75,15 @@ constructor(private val initSerEnv: Boolean,
                 "generate-node-info"
         )
 
+        private val createSchemasCmd = listOf(
```
Comment (vkolomeyko) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:78]: Why do we insist on schema migration at Network Bootstrapping stage?
All the required CorDapps may not be installed yet and some more schema changes may be necessary anyway at a later stage.

File: tools/demobench/src/main/kotlin/net/corda/demobench/model/NodeController.kt [hunk 10]
```diff
@@ -155,6 +164,23 @@ class NodeController(
...
             (networkParametersCopier ?: makeNetworkParametersCopier(config)).install(config.nodeDir)
+            @Suppress("SpreadOperator")
+            val schemaSetupCommand = jvm.commandFor(cordaPath, *schemaSetupArgs).let {
+                if (allowHibernateToManageAppSchema.value) {
+                    it + "--allow-hibernate-to-manage-app-schema"
+                } else {
+                    it
+                }
+            }.toTypedArray()
+            if (pty.runSetupProcess(schemaSetupCommand, cordaEnv, config.nodeDir.toString()) != 0) {
```
Comment (vkolomeyko) [inline tools/demobench/src/main/kotlin/net/corda/demobench/model/NodeController.kt:175]: How come that previously Demobench was working without the need to explicitly setup schema and now we have to do this additional step?

File: node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt (hunk 6 above)
Comment (blsemo) [reply node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt:55]: The node wouldn't start. I think it's in `AbstractNode`

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt (hunk 9 above)
Comment (blsemo) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkBootstrapper.kt:78]: Because we *need* to create the core schemas before we can create NodeInfos.

File: tools/demobench/src/main/kotlin/net/corda/demobench/model/NodeController.kt (hunk 10 above)
Comment (blsemo) [reply tools/demobench/src/main/kotlin/net/corda/demobench/model/NodeController.kt:175]: Because before the node would create H2 schemas via hibernate by default at first start up.

File: testing/smoke-test-utils/src/main/kotlin/net/corda/smoketesting/NodeProcess.kt (hunk 8 above)
Comment (blsemo) [reply testing/smoke-test-utils/src/main/kotlin/net/corda/smoketesting/NodeProcess.kt:151]: No - I think it would be confusing share a string array between `SmokeTests` and `DemoBench`.

**Review by vkolomeyko — DISMISSED** [review]
LGTM
Provided minor changes from #6571

**Review by vkolomeyko — APPROVED** [review]

**Comment by rick-r3** [comment]
Is it still possible to manipulate the isolation level via `dataSource.transactionIsolation` ?

**Comment by blsemo** [comment]
> 
> 
> Is it still possible to manipulate the isolation level via `dataSource.transactionIsolation` ?

I don't know, I haven't tried.
